### DISCUSSION TITLE
layers manipulation in Jupyter + legend

### DISCUSSION THREAD

**Discussion by jferencik** [visgl/deck.gl#6307] [Q&A] [answered]
Hello there,

I'd like to know if there is an easy way to interact with the deckgl widget and perform actions like  add/remove layer toggle (visibility)  in a jupyter notebook. I know of ipywidgets but it is not clear to how would such  functionality be implemented. I read some suggestions for react where setProps is reocmmended. How can i achieve the same in python?

Another questions would be related to creating legends for the current layers. Is there a standard way to do this? 

Any suggestions are welcome

**Comment by ibgreen** [comment, marked as answer]
pydeck is declarative so there is not be an imperative API for adding/removing single layers. The API underlying pydeck is set up to accept a object (dict) of all layers and views and diff it against the last dict, and thus should be possible to set a new description with a different set of layers and views.

**Reply by jferencik** [reply]
@ibgreen , thanks a lot,

do you have any example where a deck is created from a dict in pydeck?

thanks in advance

**Reply by ibgreen** [reply]
We originally designed pydeck as a lower-level component. It exposes the "raw" deck.gl layer API to Python and is great for lower level work such as composing with other Jupyter frameworks (such as streamlit etc), allows you to integrate your own custom deck.gl layers into notebooks etc. 

Also @ajduberstein has kept improving pydeck to handle common visualization requests so today it is not only a raw API but a neat basic geospatial visualization solution. 

Do note that pydeck is not the only deck.gl notebook binding. kepler.gl also has a Jupyter package and for example of an even higher level deck.gl based API approach that does offer add/remove layer, legends and much more see https://docs.unfolded.ai/jupyter/examples (which by way of disclaimer I am currently working on).

**Comment by ibgreen** [comment]
> do you have any example where a deck is created from a dict in pydeck?

Sorry, it has been quite a while since i worked on pydek. Looking closer I see that pydeck uses named parameters not a dict (presumably to provide better typing), and I do not see a way to provide a new set of parameters.

**Reply by jferencik** [reply]
This is true. There is no simple way to do this. I suspect the view state can be altered easier, however interacting with the DeckGLWidget is going  to be more complicated.

The unfolded.ai is very interesting. Do you have any spatial analysis capabilities? What about integration with databricks?

**Reply by ibgreen** [reply]
Yes here are some links. Let's keep discussions here deck.gl / pydeck focused.
- [databricks](https://docs.unfolded.ai/jupyter/cloud-notebooks/databricks) 
- [Unfolded Slack](https://unfolded.page.link/slack-signup)

**Comment by ajduberstein** [comment]
> I'd like to know if there is an easy way to interact with the deckgl widget and perform actions like add/remove layer toggle (visibility) in a jupyter notebook

This is possible by using another Jupyter widget to set the layer opacity. You can see layers being added or removed [here](https://twitter.com/ajdstein/status/1360312247562629121), for example. Check out the [Jupyter-specific features](https://pydeck.gl/jupyter.html) section of the documentation and in particular [this notebook](https://github.com/visgl/deck.gl/blob/master/bindings/pydeck/examples/05%20-%20Interacting%20with%20other%20Jupyter%20widgets.ipynb) about interacting with other widgets.

**Comment by chrisgervang** [comment]
We're discussing UI widgets for deck in #7946. Layer visibility and legends have come up as possible use cases. If this is still relevant to you it'd be great to get your ideas and feedback.
